[cpp] fix: to_if_else missing check for threshold when missing_type is not none

wuciting opened the pull request:
When missing_type is not None and fval is valid, the code before did not compare with the threshold value, resulting in a mismatch predict value in cpp code with predict value in python interface.

wuciting committed:
fix: cpp to_if_else missing check for threshold when missing_type is not none

wuciting committed:
perf: code style

jameslamb reviewed (changes requested):
Thanks for your interest in LightGBM. I don't really understand this change... can you please add a test that fails on `master` but passes here? Or at least provide a minimal, reproducible example explaining the problem?

For example:

>  in a mismatch predict value in cpp code with predict value in python interface

The Python package calls into the C API to generate predictions, so I don't understand how this statement could be true.

wuciting commented:
The original feature was to convert model param file to cpp source code (if-else format). PR: https://github.com/microsoft/LightGBM/pull/469

The bug: if the decision_type is **not** None (None is 2 in param file), 8 or 10, the tool will generate wrong cpp code, the code fails to check for thresholds if the value is not missing.

before:

A small script generated by llm to generate a model param file `lightgbm_model.txt`

```python
import numpy as np
import pandas as pd
import lightgbm as lgb
from sklearn.model_selection import train_test_split

# Generate training data
np.random.seed(42)  # Set random seed for reproducibility
size = 1000  # Increase data size to 1000 rows
num_features = 10  # Number of features

# Generate feature data, including some NaN values
X = np.random.rand(size, num_features)  # Generate 1000 rows and 10 columns of random numbers

# Randomly insert NaN values
nan_indices = np.random.choice(size * num_features, size=int(size * num_features * 0.1), replace=False)  # 10% NaN
for idx in nan_indices:
    row = idx // num_features
    col = idx % num_features
    X[row, col] = np.nan

# Generate target variable, assuming it is a linear combination plus some noise
y = 0.5 * X[:, 0] + 2 * X[:, 1] - 1.5 * X[:, 2] + np.random.normal(0, 0.1, size)  # Target variable is linearly related to features

# Convert data to pandas DataFrame
X_df = pd.DataFrame(X, columns=[f'feature{i+1}' for i in range(num_features)])
y_df = pd.Series(y)

# Split into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, test_size=0.2, random_state=42)

# Create LightGBM dataset
train_data = lgb.Dataset(X_train, label=y_train, free_raw_data=False)
test_data = lgb.Dataset(X_test, label=y_test, free_raw_data=False)

# Set LightGBM parameters
params = {
    'objective': 'regression',
    'metric': 'mse',
    'verbosity': -1,  # Turn off LightGBM output
    'num_leaves': 3,
    'learning_rate': 0.1,
    'feature_fraction': 0.9,
    'max_bin': 255,
    'min_data_in_leaf': 20,
}

# Train model
model = lgb.train(params, train_data, num_boost_round=3)
model.save_model('lightgbm_model.txt')

print("Model parameters have been exported to lightgbm_model.txt")
```

Convert it to cpp file:

```bash
./lightgbm task=convert_model input_model=lightgbm_model.txt convert_model_language=cpp convert_model=a.cpp
```

Cpp if-else file:

```cpp
#include "gbdt.h"
#include <LightGBM/utils/common.h>
#include <LightGBM/objective_function.h>
#include <LightGBM/metric.h>
#include <LightGBM/prediction_early_stop.h>
#include <ctime>
#include <sstream>
#include <chrono>
#include <string>
#include <vector>
#include <utility>
namespace LightGBM {
double PredictTree0(const double* arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr[1];if (std::isnan(fval)) {return 0.3177485457277795; } else { fval = arr[2];if (!std::isnan(fval)) {return 0.44806984060967808; } else { return 0.36627160319740354; } } }
double PredictTree0ByMap(const std::unordered_map<int, double>& arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr.count(1) > 0 ? arr.at(1) : 0.0f;if (std::isnan(fval)) {return 0.3177485457277795; } else { fval = arr.count(2) > 0 ? arr.at(2) : 0.0f;if (!std::isnan(fval)) {return 0.44806984060967808; } else { return 0.36627160319740354; } } }

double PredictTree1(const double* arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr[2];if (!std::isnan(fval)) {fval = arr[0];if (std::isnan(fval)) {return -0.036787212124237648; } else { return 0.047840481589521179; } } else { return -0.026726479674212891; } }
double PredictTree1ByMap(const std::unordered_map<int, double>& arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr.count(2) > 0 ? arr.at(2) : 0.0f;if (!std::isnan(fval)) {fval = arr.count(0) > 0 ? arr.at(0) : 0.0f;if (std::isnan(fval)) {return -0.036787212124237648; } else { return 0.047840481589521179; } } else { return -0.026726479674212891; } }

double PredictTree2(const double* arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr[1];if (std::isnan(fval)) {return -0.037347034887016677; } else { fval = arr[0];if (std::isnan(fval)) {return -0.036852757930755618; } else { return 0.048212644289095129; } } }
double PredictTree2ByMap(const std::unordered_map<int, double>& arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr.count(1) > 0 ? arr.at(1) : 0.0f;if (std::isnan(fval)) {return -0.037347034887016677; } else { fval = arr.count(0) > 0 ? arr.at(0) : 0.0f;if (std::isnan(fval)) {return -0.036852757930755618; } else { return 0.048212644289095129; } } }

double (*PredictTreePtr[])(const double*) = { PredictTree0 , PredictTree1 , PredictTree2 };

void GBDT::PredictRaw(const double* features, double *output, const PredictionEarlyStopInstance* early_stop) const {
        int early_stop_round_counter = 0;
        std::memset(output, 0, sizeof(double) * num_tree_per_iteration_);
        for (int i = 0; i < num_iteration_for_pred_; ++i) {
                for (int k = 0; k < num_tree_per_iteration_; ++k) {
                        output[k] += (*PredictTreePtr[i * num_tree_per_iteration_ + k])(features);
                }
                ++early_stop_round_counter;
                if (early_stop->round_period == early_stop_round_counter) {
                        if (early_stop->callback_function(output, num_tree_per_iteration_))
                                return;
                        early_stop_round_counter = 0;
                }
        }
}

double (*PredictTreeByMapPtr[])(const std::unordered_map<int, double>&) = { PredictTree0ByMap , PredictTree1ByMap , PredictTree2ByMap };

void GBDT::PredictRawByMap(const std::unordered_map<int, double>& features, double* output, const PredictionEarlyStopInstance* early_stop) const {
        int early_stop_round_counter = 0;
        std::memset(output, 0, sizeof(double) * num_tree_per_iteration_);
        for (int i = 0; i < num_iteration_for_pred_; ++i) {
                for (int k = 0; k < num_tree_per_iteration_; ++k) {
                        output[k] += (*PredictTreeByMapPtr[i * num_tree_per_iteration_ + k])(features);
                }
                ++early_stop_round_counter;
                if (early_stop->round_period == early_stop_round_counter) {
                        if (early_stop->callback_function(output, num_tree_per_iteration_))
                                return;
                        early_stop_round_counter = 0;
                }
        }
}

void GBDT::Predict(const double* features, double *output, const PredictionEarlyStopInstance* early_stop) const {
        PredictRaw(features, output, early_stop);
        if (average_output_) {
                for (int k = 0; k < num_tree_per_iteration_; ++k) {
                        output[k] /= num_iteration_for_pred_;
                }
        }
        if (objective_function_ != nullptr) {
                objective_function_->ConvertOutput(output, output);
        }
}

void GBDT::PredictByMap(const std::unordered_map<int, double>& features, double* output, const PredictionEarlyStopInstance* early_stop) const {
        PredictRawByMap(features, output, early_stop);
        if (average_output_) {
                for (int k = 0; k < num_tree_per_iteration_; ++k) {
                        output[k] /= num_iteration_for_pred_;
                }
        }
        if (objective_function_ != nullptr) {
                objective_function_->ConvertOutput(output, output);
        }
}

double PredictTree0Leaf(const double* arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr[1];if (std::isnan(fval)) {return 0; } else { fval = arr[2];if (!std::isnan(fval)) {return 1; } else { return 2; } } }
double PredictTree0LeafByMap(const std::unordered_map<int, double>& arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr.count(1) > 0 ? arr.at(1) : 0.0f;if (std::isnan(fval)) {return 0; } else { fval = arr.count(2) > 0 ? arr.at(2) : 0.0f;if (!std::isnan(fval)) {return 1; } else { return 2; } } }

double PredictTree1Leaf(const double* arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr[2];if (!std::isnan(fval)) {fval = arr[0];if (std::isnan(fval)) {return 0; } else { return 2; } } else { return 1; } }
double PredictTree1LeafByMap(const std::unordered_map<int, double>& arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr.count(2) > 0 ? arr.at(2) : 0.0f;if (!std::isnan(fval)) {fval = arr.count(0) > 0 ? arr.at(0) : 0.0f;if (std::isnan(fval)) {return 0; } else { return 2; } } else { return 1; } }

double PredictTree2Leaf(const double* arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr[1];if (std::isnan(fval)) {return 0; } else { fval = arr[0];if (std::isnan(fval)) {return 1; } else { return 2; } } }
double PredictTree2LeafByMap(const std::unordered_map<int, double>& arr) { const std::vector<uint32_t> cat_threshold = {};double fval = 0.0f; fval = arr.count(1) > 0 ? arr.at(1) : 0.0f;if (std::isnan(fval)) {return 0; } else { fval = arr.count(0) > 0 ? arr.at(0) : 0.0f;if (std::isnan(fval)) {return 1; } else { return 2; } } }

double (*PredictTreeLeafPtr[])(const double*) = { PredictTree0Leaf , PredictTree1Leaf , PredictTree2Leaf };

void GBDT::PredictLeafIndex(const double* features, double *output) const {
        int total_tree = num_iteration_for_pred_ * num_tree_per_iteration_;
        for (int i = 0; i < total_tree; ++i) {
                output[i] = (*PredictTreeLeafPtr[i])(features);
        }
}
double (*PredictTreeLeafByMapPtr[])(const std::unordered_map<int, double>&) = { PredictTree0LeafByMap , PredictTree1LeafByMap , PredictTree2LeafByMap };

void GBDT::PredictLeafIndexByMap(const std::unordered_map<int, double>& features, double* output) const {
        int total_tree = num_iteration_for_pred_ * num_tree_per_iteration_;
        for (int i = 0; i < total_tree; ++i) {
                output[i] = (*PredictTreeLeafByMapPtr[i])(features);
        }
}
}  // namespace LightGBM
```

wuciting committed:
Merge branch 'master' into master

StrikerRUS commented:
Kindly ping @wuciting for a requested test.

We already have `if-else` test at our CI: 
https://github.com/microsoft/LightGBM/blob/18c11f861118aa889b9d4579c2888d5c908fd250/.ci/test.sh#L71-L83

I think it's possible to add a check for correctly generated `LightGBM_model.txt` here https://github.com/microsoft/LightGBM/blob/master/tests/cpp_tests/test.py.

wuciting committed:
Merge branch 'microsoft:master' into master

wuciting committed:
add test

wuciting commented:
I added a test for if-else with nan input. I also fixed the original if-else test. It did not compile the executable after the cpp file is changed, so it always produce the same prediction result.

wuciting commented:
> @wuciting please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

wuciting committed:
fix: test

wuciting committed:
Merge branch 'microsoft:master' into master
